File: 프로그래머스/2/181187. 두 원 사이의 정수 쌍/두 원 사이의 정수 쌍.cpp
#include <string>
#include <vector>
#include <cmath>

using namespace std;

long long solution(int r1, int r2) {
    long long answer = 0;
    
    long long R1 = 1LL * r1 * r1;
    long long R2 = 1LL * r2 * r2;
    for (int x = 1; x <= r2; x++) {
        long long X = 1LL * x * x;
        
        long long max_v = floor(sqrt(R2 - X));
        long long min_v = 0;
        if (x <= r1) {
            min_v = ceil(sqrt(R1 - X));
        }
        
        answer += (max_v - min_v) + 1;
    }
    return answer * 4;
}

File: 프로그래머스/2/181187. 두 원 사이의 정수 쌍/두 원 사이의 정수 쌍.java
class Solution {
    public long solution(int r1, int r2) {
        long answer = 0;
        
        // r1^2 <= x^2 + y^2 <= r2^2
        // sqrt(r1^2 - x^2) <= y <= sqrt(r2^2 - x^2)
        for (int x = 1; x <= r2; x++) {
            
            long maxY = (long) Math.floor(Math.sqrt((long) r2*r2 - (long) x*x));
            long minY = (long) Math.ceil(Math.sqrt((long) r1*r1 - (long) x*x));
            answer += maxY - minY + 1;
        }
        
        return answer * 4;
    }
}